src/tracing/trace_flamegraph.c

```diff
@@ -0,0 +1,263 @@
+   +----------------------------------------------------------------------+
+   | Copyright (c) 2002-2020 Derick Rethans                               |
+   +----------------------------------------------------------------------+
+   | This source file is subject to version 1.01 of the Xdebug license,   |
+   | that is bundled with this package in the file LICENSE, and is        |
+   | available at through the world-wide-web at                           |
+   | https://xdebug.org/license.php                                       |
+   | If you did not receive a copy of the Xdebug license and are unable   |
+   | to obtain it through the world-wide-web, please send a note to       |
+   | derick@xdebug.org so we can mail you a copy immediately.             |
+   +----------------------------------------------------------------------+
+ */
+#include "php.h"
+#include "ext/standard/php_string.h"
+
+#include "php_xdebug.h"
+#include "tracing_private.h"
+#include "trace_flamegraph.h"
+
+#include "lib/lib_private.h"
+#include "lib/var_export_line.h"
+
+extern ZEND_DECLARE_MODULE_GLOBALS(xdebug);
+
+flamegraph_stack *fg_stack_create()
+{
+	flamegraph_stack *ret;
+	ret = xdmalloc(sizeof(flamegraph_stack));
+	ret->head = NULL;
+	return ret;
+}
+
+void fg_stack_free(flamegraph_stack *stack)
+{
+	flamegraph_stack_item *cur, *prev;
+	cur = stack->head;
+	while (cur) {
+		prev = cur->next;
+		xdfree(cur->prefix);
+		xdfree(cur);
+		cur = prev;
+	}
+	xdfree(stack);
+}
+
+static inline void fg_stack_inc(flamegraph_stack *stack, int inc)
+{
+	if (NULL == stack->head) {
+		return;
+	}
+	stack->head->value += inc;
+}
+
+static inline void fg_stack_push(flamegraph_stack *stack, char *prefix)
+{
+	flamegraph_stack_item *item;
+	item = xdmalloc(sizeof(flamegraph_stack_item));
+	item->prefix = xdstrdup(prefix);
+	item->value = 0;
+	item->next = stack->head;
+	stack->head = item;
+}
+
+static inline flamegraph_stack_item *fg_stack_get_head(flamegraph_stack *stack)
+{
+	return stack->head;
+}
+
+static inline flamegraph_stack_item *fg_stack_pop(flamegraph_stack *stack)
+{
+	flamegraph_stack_item *item;
+	if (NULL == stack->head) {
+		return 0;
+	}
+	item = stack->head;
+	stack->head = item->next;
+	return item;
+}
+
+static inline int compute_inclusive_value(xdebug_trace_flamegraph_context *context, function_stack_entry *fse)
+{
+	int value = 0, current_mem;
+
+	switch (context->mode) {
+		case XDEBUG_TRACE_OPTION_FLAMEGRAPH_MEM:
+			/* We compare with 'memory' because 'prev_memory' is not memory when
+			   starting the function execution, 'memory' is. */
+			current_mem = zend_memory_usage(0);
+			if (current_mem < fse->memory) {
+				/* When memory is below 0, flamegraph generator will error, and
+				   you won't have a good visual.
+				   This happens when garbage collection happened during this
+				   function and freed something it didn't allocate, I guess. */
+				value = 0;
+			} else {
+				value = current_mem - fse->memory;
+			}
+			break;
```

Comment:
Yes I agree, but as a matter of fact, I wrote this because I do use it, it's imprecise, but it gives serious hints for finding potential leaks. I don't know any other PHP memory visualisation consumption tools and got used to this method, even if wrong. Flamegraph do have memory oriented graphs, but I don't master them yet, it's barely documented how to write the output file correctly.

I'd like to keep the memory one, documented as being imprecise and experimental, but if it's really blocking for you, I'll remove it if that's a condition for having the time based flamegraph.

Reply:
From https://www.brendangregg.com/FlameGraphs/memoryflamegraphs.html which describe more many methods for analyse the memory allocations, it seems that the first two are malloc based counts:

 - first one simply count the malloc() calls, this mean that in our trace output, the final number is an absolute malloc() call number; we cannot do that in tracing handler because it seems we can't really hook on memory allocation (can xdebug do that in any other way ?)
 - second one will count the aggregate sum of size_t being allocated by malloc calls, the number in our trace output will be then total memory that has been allocated (in my algorithm, it's the total memory that has been allocated *but not freed*).

It seems that all described methods for generated memory flamegraph have pros and cons, there is no golden hammer.

Is there a pragmatic reason why you don't want the memory trace as I implemented it ? Do you have suggestions how to improve it ? Are you afraid that possible allocations from xdebug themselves might alter the result ?